File: java/Reflection.java
import java.lang.reflect.*;

public class Reflection {

    static class Person {
        public String name;
        private int age;

        Person(String name, int age) {
            this.name = name;
            this.age = age;
        }
        public String greet() { return "Hi, I'm " + name; }
        private int getAge() { return age; }

        @Override public String toString() { return name + "/" + age; }
    }

    public static void main(String[] args) throws Exception {
        Class<?> c = Person.class;

        // READ
        for (Field f : c.getDeclaredFields()) System.out.println("field: " + f);
        for (Method m : c.getDeclaredMethods()) System.out.println("method: " + m);
        for (Constructor<?> k : c.getDeclaredConstructors()) System.out.println("contructor: " + k);

        Person p = new Person("Alice", 30);
        for (Field f : c.getDeclaredFields()) {
            f.setAccessible(true);
            System.out.println(f.getName() + "=" + f.get(p));
        }

        // INVOKE
        System.out.println(c.getMethod("greet").invoke(p));

        Method getAge = c.getDeclaredMethod("getAge");
        getAge.setAccessible(true);
        System.out.println(getAge.invoke(p));

        Person p2 = (Person) c.getDeclaredConstructor(String.class, int.class).newInstance("Bob", 25);
        System.out.println(p2);

        // MODIFY
        System.out.println("before: " + p);
        c.getDeclaredField("name").set(p, "Alice2");
        Field age = c.getDeclaredField("age");
        age.setAccessible(true);
        age.set(p, 31);
        System.out.println("after: " + p);
    }
}

File: scala/Reflection.scala
// --- REFLECTION (Java reflection in Scala 3) ---

case class Person(name: String, age: Int):
  def greet(): String = s"Hi, I'm $name"

class Counter(var count: Int)

@main def run(): Unit =
  val p = Person("Alice", 30)
  val cls = p.getClass

  println("Fields:")
  cls.getDeclaredFields.foreach { f =>
    f.setAccessible(true)
    println(s"  ${f.getName}: ${f.getType.getSimpleName} = ${f.get(p)}")
  }

  println("\nMethods:")
  cls.getDeclaredMethods
    .filterNot(m => m.isSynthetic || m.getName.contains("$"))
    .foreach(m => println(s"  ${m.getName}(${m.getParameterTypes.map(_.getSimpleName).mkString(", ")}): ${m.getReturnType.getSimpleName}"))

  println("\nInvoking 'greet' dynamically:")
  val greetMethod = cls.getDeclaredMethod("greet")
  println(s"  ${greetMethod.invoke(p)}")

  println("\nMutating 'count' field on Counter:")
  val c = Counter(0)
  val countField = c.getClass.getDeclaredField("count")
  countField.setAccessible(true)
  countField.set(c, 42)
  println(s"  count after set(42) = ${countField.get(c)}")

  println("\nInstantiating Person via Class.forName:")
  val dynClass = Class.forName(cls.getName)
  val ctor = dynClass.getDeclaredConstructors.head
  val dynPerson = ctor.newInstance("Bob", Int.box(25))
  println(s"  created: $dynPerson")
